Clarify `root_motion_track` usage and parent node

scotmcp opened the pull request:
clarified root_motion_track usage

<snip
The path to the Animation track used for root motion. Paths must be valid scene-tree paths to a node, and must be specified starting from the parent node of the node that will reproduce the animation.
/snip>

I added this line immediately after the one above:	The parent node is [i]usually[/i] the [member AnimationMixer.root_node].


<!--
Please target the `master` branch in priority.

Relevant fixes are cherry-picked for stable branches as needed by maintainers.

To speed up the contribution process and avoid CI errors, please set up pre-commit hooks locally:
https://docs.godotengine.org/en/latest/contributing/development/code_style_guidelines.html
-->

m4gr3d committed:
Fix ANRs reported by the Google Play Console

- Add support for dispatching input on the render thread (UI thread is the current default) when `input_buffering` and `accumulated_input` are disabled. At the expense of latency, this helps prevent 'heavy' applications / games from blocking the UI thread (the default behavior)  which may cause the application to ANR.

- Remove GLSurfaceView logic causing the UI thread to wait on the GL thread during lifecycle events. The removed logic would cause the UI thread to ANR when the GL thread is blocked.

rune-scape committed:
GDScript: Invalidate cached parser chain when reloading

mihe committed:
Add input event callback to `DisplayServerHeadless`

RandomShaper committed:
Enable usage of asan-enabled versions of prebuilt libraries

dsnopek committed:
Fix building with `use_volk=yes` on MacOS

CookieBadger committed:
fix animation bezier crash on undo

oshman99 committed:
NavigationServer3D.map_get_closest_point_to_segment - add an additional shortest distance check

For a case when shortest distance is between some point located on a face's
edge and some point located on a line segment.

Gurvan committed:
Use Hermite instead of Bezier for glTF spline interpolation

KoBeWi committed:
Prevent selecting when a CanvasItem is selected

RandomShaper committed:
D3D12: Use the right state for resources in certain heap types

bruvzg committed:
[Button] Use align_to_largest_stylebox for min. size calculation.

MewPurPur committed:
Fix unnecessary overbright modulates used in a few editors

KoBeWi committed:
Speed up scene group scanning for text scenes

passivestar committed:
Make the frame time and info boxes have same margins

markdibarry committed:
Add alternative pixel rounding

poga committed:
Fix: set max_width for icons in the quick open popup

Co-authored-by: Tomek <kobewi4e@gmail.com>

AThousandShips committed:
[Tests] Fix unit tests in template builds

KoBeWi committed:
Fix some TileMapLayer editing problems

adamscott committed:
Add `bigint` support on JS value conversion

Geometror committed:
[GraphEdit] Add a note about the behavior of get_children()

scotmcp committed:
Update AnimationMixer.xml

clarified root_motion_track usage

AThousandShips commented:
Please update the commit PR title to have a more descriptive message

scotmcp commented:
Updated as requested

AThousandShips commented:
Meant the title, not the description, the part that says "Update AnimationMixer.xml"

Mickeon commented:
What this PR does as of writing this is just add the following line:
```
The parent node is [i]usually[/i] the [member AnimationMixer.root_node].
```
Not sure about this clarification. If anything, I am more confused than before. In what situation is the parent node _not_ the root node, or why does that matter?

By the way, because it's part of the same class reference you may omit the name of the class (`[member root_node]`).

adamscott committed:
Fix assignations to non-existing keys and clean-up

scotmcp commented:
> What this PR does as of writing this is just add the following line:
> 
> ```
> The parent node is [i]usually[/i] the [member AnimationMixer.root_node].
> ```
> 
> Not sure about this clarification. If anything, I am more confused than before. In what situation is the parent node _not_ the root node, or why does that matter?
> 
> By the way, because it's part of the same class reference you may omit the name of the class (`[member AnimationMixer.root_node]` -> `[member root_node]`).

Well, it doesn't HAVE to be the root node, but it probably is. What is confusing about the original wording is "starting from the parent node of the node that will reproduce the animation." What is the node that reproduce the animation? Is that the skeleton, or the animation player? Does that mean the animation player's node, and then you nodepath all the way to the skeleton, or is it the parent node of the skeleton?  This little clarification might be what a NEW developer needs to understand the sentence that comes before it.

In my case I am instantiated scenes with animationplayer and animationtree, and the skeleton is on the parent scene that instantiated it, so I have to assign root_node and root_motion_track through code (you are unable assign it in the editor), it took me quite a bit of playing around with paths by using NodePaths that that sentence could have meant from my point of view having never done this specific thing before. Now that I know, I know, but this will help someone who does not already know.

scotmcp committed:
removed class from member markup reference

scotmcp commented:
I removed the class from the marked up  member reference

Trinovantes committed:
Fix asset pipeline coordinate systems doc url

ashish0kumar committed:
Fix "icon" parameter in EditorPlugin.add_custom_type is not optional, but docs says it is# This is a combination of 2 commits.

Fix "icon" parameter in EditorPlugin.add_custom_type is not optional, but docs says it is

Update doc/classes/EditorPlugin.xml

Co-authored-by: A Thousand Ships <96648715+AThousandShips@users.noreply.github.com>

TokageItLab committed:
Fix Bezier track init value on caching

kitbdev committed:
Clarify word separator docs about shortcuts

passivestar committed:
[macOS] Change the distraction-free mode shortcut

Chaosus committed:
Prevent shader crash when doing invalid operation on boolean vector

AdriaandeJongh committed:
fixed the y-sorting description having higher and lower inverted (oops!)

akien-mga committed:
Revert "Make freed object different than null in comparison operators"

This reverts commit 150b50cfcde95f74419ade2811d963224e96fc98.

As discussed with the GDScript team, this has some implications which aren't
fully consensual yet, and which we want to revisit.

For now we revert to the 4.2 behavior for the 4.3 release, to avoid breaking
user expectations.

ashish0kumar committed:
Fix Info about Global library on add_animation_library method in doc/classes/AnimationMixer.xml

Fix Info about Global library on add_animation_library method in doc/classes/AnimationMixer.xml

Fix Info about Global library on add_animation_library method

Co-authored-by: Silc Lizard (Tokage) Renew <tokage.it.lab@gmail.com>

allenwp committed:
Updated documentation for new EditorExportPlugin behaviour.
This behaviour was introduced in #90365. This also fixes some inconsistencies in the docs and adds clarification of how the _export_file function works.

kleonc committed:
Fix AtlasTexture::draw_rect flipping for non-zero margin

jsjtxietian committed:
Fix sprite frame editor crash with null `frames` pounter

Chaosus committed:
Fix orphan StringName's in ShaderLanguage

KoBeWi committed:
Fix RMB erasing of tiles

Faless committed:
[Web] Remove unnecessary EMCC_FORCE_STDLIBS in dlink builds

As discussed with upstream, the C/C++ standard library is always fully
included when building with MAIN_MODULE=1, so using EMCC_FORCE_STDLIBS
is not necessary in our case.

lawnjelly committed:
Physics Interpolation - Fix `interpolated_transform_2d`

Uses the skew correct `Transform2D::interpolate_with()` function rather than the bugged 3.x version.

ayanchavand committed:
Fix: AnimationMixer Example Code

timothyqiu committed:
Update Node dock when theme changes

MikeXYZW committed:
Clarify that the `Mesh.ARRAY_NORMAL` array will normalize its contents internally

m4gr3d committed:
Fix the issue causing the logo to not show when using the `compatibility` renderer

hakro committed:
Add : to node properties, to differentiate them from node paths

KoBeWi committed:
Fix undoredo handling in some dialogs

JacobMillner committed:
fix the sorting of blend_times in animation player

Modify BlendKey's sort to use  AlphaCompare in order to create a deterministic sort

Co-authored-by: A Thousand Ships <96648715+AThousandShips@users.noreply.github.com>

Geometror committed:
Fix `GraphFrame` autoshrink with high DPI displays/custom UI scales

paulsinnett committed:
Fix preloading a zip in the web editor

JustVicctor committed:
Fix custom resources often missing from Quick Load dialog

zeux committed:
Fix AABB computation for position compression to not depend on vertex order

The previous computation was dependent on the vertex order in two ways:

- If the first vertex was on the AABB boundary, the AABB would be
  increased by the epsilon due to size clamping
- Every time the AABB would get expanded, we would recompute end from
  size and reconstruct size again, which resulted in slow floating point
  drift.

In isolation this may not seem like it matters, but it means that the
same mesh with a different vertex order may result in a slightly different
AABB. This can be a significant issue due to shadow meshes and their use in
depth prepass: shadow meshes reorder vertex data as part of the
deduplication process, as they append one unique position at a time and
as such remove the duplicate positions; this can result in a different
AABB which would result in a different reconstructed vertex position
during a depth pre-pass, causing mesh self-occlusion.

Hilderin committed:
Fix adding a translation CSV results in errors on initial import for many types of resources

scotmcp commented:
Hey guys, any updated thoughts on this? I really believe this is helpful for people who don't already know. It's just a clarifying statement to help new people who need to assign root_motion_track through code understand which node is most likely the one that needs to be pathed.

m4gr3d committed:
Fix crashes reported by the Google Play Console

marcospb19 committed:
Mention timeout unit in ENetConnection::service docs

Arnklit committed:
Clamp bezier handle length to half the length of animation

TokageItLab reviewed (changes requested):
Since it applies to the whole Animation that the track path is relative to the root_node, not only the root_motion_track, you should fill in the details in `track_set_path()` document, then quote it in root_motion_track document as "See also [method Animation.track_set_path]."

bruvzg committed:
[MoltenVK] Fix downscaled hiDPI window pixelation.

scotmcp commented:
I am not sure I understand, AnimationPlayer and AnimationTree do not inherit from Animation...The Animation class methods seem to be to more about manipulating an animation object and all the methods are entirely about setting properties of an animation, not about playing or reproducing them.

track_set_path() - is a method that identifies the track the developer wants to manipulate in their script.

root_motion_track - is a property set so the AnimationPlayer Node and AnimationTree Node knows where to find the root motion track.

I don't think they are equivalent or even all that similar.

Perhaps the same sentence should be included in that method description as well, AND a  "See also "  but really I believed think it should be referenced as only a See also quote as that would imply they are inter-related, and I don't think they are.

TokageItLab commented:
You should read the `track_set_path` documentation. They say the same thing about setting trackpaths.

scotmcp commented:
You should re-read what I said.....maybe they should say the same things, but they are not really related nor dependent on one another.

TokageItLab commented on doc/classes/Animation.xml:
```diff
@@ -561,7 +561,7 @@
 			<description>
 				Sets the path of a track. Paths must be valid scene-tree paths to a node and must be specified starting from the parent node of the node that will reproduce the animation. Tracks that control properties or bones must append their name after the path, separated by [code]":"[/code].
-				For example, [code]"character/skeleton:ankle"[/code] or [code]"character/mesh:transform/local"[/code].
+				For example, [code]"character/skeleton:ankle"[/code] or [code]"character/mesh:transform/local"[/code]. See also [member AnimationMixer.root_motion_track].
```
The reference is opposite because root_motion_track depends on the format of the Track in Animation.

scotmcp replied:
I cross referenced it both directions

TokageItLab replied:
Since the same document is not needed in two places, only one link is sufficient. Also unnecessary links confuse people who are not interested in root_motion but came to read the animation documentation.

scotmcp replied:
Then please cherry pick the parts that you want.

scotmcp commented:
I cross referenced it both directions

TokageItLab commented on doc/classes/AnimationMixer.xml:
```diff
@@ -304,8 +304,9 @@
-			The path to the Animation track used for root motion. Paths must be valid scene-tree paths to a node, and must be specified starting from the parent node of the node that will reproduce the animation. To specify a track that controls properties or bones, append its name after the path, separated by [code]":"[/code]. For example, [code]"character/skeleton:ankle"[/code] or [code]"character/mesh: ...
-			If the track has type [constant Animation.TYPE_POSITION_3D], [constant Animation.TYPE_ROTATION_3D] or [constant Animation.TYPE_SCALE_3D] the transformation will be canceled visually, and the animation will appear to stay in place. See also [method get_root_motion_position], [method get_root_motion_rotation], [method get_root_motion_scale] and [RootMotionView].
+			The path to the Animation track used for root motion. Paths must be valid scene-tree paths to a node, and must be specified starting from the parent node of the node that will reproduce the animation.
+			The parent node is [i]usually[/i] the [member root_node]. To specify a track that controls properties or bones, append its name after the path, separated by [code]":"[/code]. For example, [code]"character/skeleton:ankle"[/code] or [code]"character/mesh:transform/local"[/code]. See also [method Animation.track_set_path].
```
```suggestion
			The [member root_motion_track] uses the same format as [method Animation.track_set_path], but note that a bone must be specified.
```

TokageItLab commented on doc/classes/Animation.xml:
```diff
@@ -561,7 +561,7 @@
 			<param index="1" name="path" type="NodePath" />
 			<description>
 				Sets the path of a track. Paths must be valid scene-tree paths to a node and must be specified starting from the parent node of the node that will reproduce the animation. Tracks that control properties or bones must append their name after the path, separated by [code]":"[/code].
```
```suggestion
				Sets the path of a track. Paths must be valid scene-tree paths to a node and must be specified starting from the [member AnimationMixer.root_node]. that will reproduce the animation. Tracks that control properties or bones must append their name after the path, separated by [code]":"[/code].
```
If the root node is mentioned, it should be here as @lyuma suggested.

scotmcp replied:
That's what the first version of my commit said!!!?!!!

TokageItLab replied:
That was inappropriate because it was written in the root_motion_track location.

scotmcp replied:
you just asked me to update my PR to clarifying that the parent node was the root_node... that's what my first commit said.

This doesn't need to be this difficult....you have to understand people read these docs because they don't know....you are reviewing them as people who already know. This isn't documenting the class's source code, it's documenting how to use the class. What I wrote was correct and accurate, and you chased me around the bush back to the beginning, and asked me to change it basically almost word for word what I first Pull Requested...I am sorry, but this is frustrating.

TokageItLab replied:
Your first commit is this, which is quite different from my suggestion where it describes root_node.
https://github.com/godotengine/godot/pull/93752/commits/fbfddff90eac6e6572bd9bbd12bde31673b2bc2a

scotmcp replied:
you mean this part "The parent node is [i]usually[/i] the [member root_node]." ?

scotmcp replied:
Oh I see my error....my apoligies

TokageItLab commented on doc/classes/Animation.xml:
```diff
@@ -560,8 +560,8 @@
-				Sets the path of a track. Paths must be valid scene-tree paths to a node and must be specified starting from the parent node of the node that will reproduce the animation. Tracks that control properties or bones must append their name after the path, separated by [code]":"[/code].
-				For example, [code]"character/skeleton:ankle"[/code] or [code]"character/mesh:transform/local"[/code].
+				Sets the path of a track. Paths must be valid scene-tree paths to a node and must be specified starting from the [member AnimationMixer.root_node] that will reproduce the animation. Tracks that control properties or bones must append their name after the path, separated by [code]":"[/code].
+				For example, [code]"character/skeleton:ankle"[/code] or [code]"character/mesh:transform/local"[/code]. See also [member AnimationMixer.root_motion_track].
```
```suggestion
				For example, [code]"character/skeleton:ankle"[/code] or [code]"character/mesh:transform/local"[/code].
```

TokageItLab commented:
After completing the changes, please rebase them to follow the pull request workflow.
https://docs.godotengine.org/it/4.x/contributing/workflow/pr_workflow.html#the-interactive-rebase

AThousandShips commented on doc/classes/AnimationMixer.xml:
```diff
@@ -304,8 +304,9 @@
-			If the track has type [constant Animation.TYPE_POSITION_3D], [constant Animation.TYPE_ROTATION_3D] or [constant Animation.TYPE_SCALE_3D] the transformation will be canceled visually, and the animation will appear to stay in place. See also [method get_root_motion_position], [method get_root_motion_rotation], [method get_root_motion_scale] and [RootMotionView].
+			The path to the Animation track used for root motion. Paths must be valid scene-tree paths to a node, and must be specified starting from the parent node of the node that will reproduce the animation.
+			The [member root_motion_track] uses the same format as [method Animation.track_set_path], but note that a bone must be specified.
+			If the track has type [constant Animation.TYPE_POSITION_3D], [constant Animation.TYPE_ROTATION_3D] or [constant Animation.TYPE_SCALE_3D] the transformation will be canceled visually, and the animation will appear to stay in place. See also [method get_root_motion_position], [method get_root_motion_rotation], [method get_root_motion_scale] and [RootMotionView]. 
```
```suggestion
			If the track has type [constant Animation.TYPE_POSITION_3D], [constant Animation.TYPE_ROTATION_3D] or [constant Animation.TYPE_SCALE_3D] the transformation will be canceled visually, and the animation will appear to stay in place. See also [method get_root_motion_position], [method get_root_motion_rotation], [method get_root_motion_scale], and [RootMotionView]. 
```
Please use [Oxford comma](https://docs.godotengine.org/en/latest/contributing/documentation/docs_writing_guidelines.html#use-the-oxford-comma-to-enumerate-anything).

AThousandShips commented on doc/classes/AnimationMixer.xml:
```diff
@@ -304,8 +304,9 @@
-			If the track has type [constant Animation.TYPE_POSITION_3D], [constant Animation.TYPE_ROTATION_3D] or [constant Animation.TYPE_SCALE_3D] the transformation will be canceled visually, and the animation will appear to stay in place. See also [method get_root_motion_position], [method get_root_motion_rotation], [method get_root_motion_scale] and [RootMotionView].
+			The path to the Animation track used for root motion. Paths must be valid scene-tree paths to a node, and must be specified starting from the parent node of the node that will reproduce the animation.
+			The [member root_motion_track] uses the same format as [method Animation.track_set_path], but note that a bone must be specified.
+			If the track has type [constant Animation.TYPE_POSITION_3D], [constant Animation.TYPE_ROTATION_3D] or [constant Animation.TYPE_SCALE_3D] the transformation will be canceled visually, and the animation will appear to stay in place. See also [method get_root_motion_position], [method get_root_motion_rotation], [method get_root_motion_scale], and [RootMotionView]. 
```
```suggestion
			If the track has type [constant Animation.TYPE_POSITION_3D], [constant Animation.TYPE_ROTATION_3D], or [constant Animation.TYPE_SCALE_3D] the transformation will be canceled visually, and the animation will appear to stay in place. See also [method get_root_motion_position], [method get_root_motion_rotation], [method get_root_motion_scale], and [RootMotionView].
```

AThousandShips commented:
Please squash your commits into one, see [here](https://docs.godotengine.org/en/latest/contributing/workflow/pr_workflow.html#the-interactive-rebase)

scotmcp commented:
> Please squash your commits into one, see [here](https://docs.godotengine.org/en/latest/contributing/workflow/pr_workflow.html#the-interactive-rebase)

It's throwing an error

sh-5.2$ git rebase -i upstream/master
fatal: invalid upstream 'upstream/master'

scotmcp commented:
I was able to rebase it locally, there were some conflicts I had to resolve, however that command still results in fatal error

scotmcp commented:
Just checking because of the timing, did you see my previous comment that I was able to rebase on my fork?

AThousandShips commented:
You used merge commits so doesn't look like a correct rebase, but can't tell directly

scotmcp commented:
I appreciate your patience, and sorry for the misunderstand...two similar classes with similar edits.

AThousandShips commented:
No worries, but there are no conflicts so you just need to squash and force push and it'll work out fine

scotmcp commented:
So, I did a push but it didn't do anything

sh-5.2$ git push --force origin origin/patch-1

Total 0 (delta 0), reused 0 (delta 0), pack-reused 0 (from 0)

AThousandShips commented:
Try `git push --force origin patch-1`, you would have pushed to the `origin/patch-1` branch on `origin`

scotmcp commented:
sh-5.2$ git status
HEAD detached from fbfddff90e
nothing to commit, working tree clean

scotmcp commented:
git push --force just says, "you are not currently on a branch"

AThousandShips commented:
That's why, you're detached, you need to do `git checkout patch-1`

scotmcp commented:
Well now is just says everything is up to date.

i am about to just make a patch, delete my fork and just patch the fork and make a new PR...are you ok with that?

AThousandShips commented:
You should rebase again, your rebase failed, so just try a rebase again

scotmcp committed:
Added reference to AnimationMixer.root_motion_track

Added reference to AnimationMixer.root_motion_track
Added reference to Animation.track_set_path from AnimationMixer.root_motion_track

Added reference to Animation.track_set_path from AnimationMixer.root_motion_track
Refer to AnimationMixer.root Node from Animation.track_set_path

Co-authored-by: Silc Lizard (Tokage) Renew <tokage.it.lab@gmail.com>
Update usage guidelines for [AnimationMixer.root_motion_track]

Co-authored-by: Silc Lizard (Tokage) Renew <tokage.it.lab@gmail.com>
Remove Reference from Animation class

Co-authored-by: Silc Lizard (Tokage) Renew <tokage.it.lab@gmail.com>
Added a Comma

Co-authored-by: A Thousand Ships <96648715+AThousandShips@users.noreply.github.com>
Added another comma

Co-authored-by: A Thousand Ships <96648715+AThousandShips@users.noreply.github.com>

AThousandShips commented:
You still have 4 commits in your branch

scotmcp commented:
NVM, I got it now....good lord

scotmcp commented:
It should be in good shape now, all squshed and pushed

AThousandShips commented:
No you still have 4 commits in your branch, if you need help I can fix your branch tomorrow for you

scotmcp commented:
![image](https://github.com/godotengine/godot/assets/79280460/6a7c4e83-c178-4d37-9739-368bec9bd930)
it looks looks right to me, I have other branches from this, but commit 8d6ce which is the head and master of patch-1 should be alright

AThousandShips commented:
Look at the top of the PR page, it's 4, don't trust your local tool fully

scotmcp commented:
Forget it, I made a patch file, I am going to do a fresh fork. I don't have anything worth saving

Thank you...I will do a fresh PR when I am done and it will be clean.
